src/main/scala/org/broadinstitute/dsde/workbench/leonardo/auth/IamProxyAuthProvider.scala

```diff
@@ -0,0 +1,143 @@
+package org.broadinstitute.dsde.workbench.leonardo.auth
+
+import java.util.concurrent.TimeUnit
+
+import cats.implicits._
+import akka.actor.ActorSystem
+import akka.http.scaladsl.model.headers.OAuth2BearerToken
+import com.google.common.cache.{CacheBuilder, CacheLoader}
+import com.typesafe.config.Config
+import net.ceedubs.ficus.Ficus._
+import org.broadinstitute.dsde.workbench.google.GoogleCredentialModes.Token
+import org.broadinstitute.dsde.workbench.google.{GoogleIamDAO, HttpGoogleIamDAO}
+import org.broadinstitute.dsde.workbench.leonardo.auth.IamProxyAuthProvider.{CacheKey, ProjectAuthCacheKey}
+import org.broadinstitute.dsde.workbench.leonardo.model.NotebookClusterActions.NotebookClusterAction
+import org.broadinstitute.dsde.workbench.leonardo.model.ProjectActions.ProjectAction
+import org.broadinstitute.dsde.workbench.leonardo.model.google.ClusterName
+import org.broadinstitute.dsde.workbench.leonardo.model.{LeoAuthProvider, ServiceAccountProvider}
+import org.broadinstitute.dsde.workbench.model.{UserInfo, WorkbenchEmail}
+import org.broadinstitute.dsde.workbench.model.google.{GoogleProject, IamPermission}
+
+import scala.concurrent.duration.{FiniteDuration, MINUTES}
+import scala.concurrent.{ExecutionContext, Future}
+
+object IamProxyAuthProvider {
+  private sealed trait CacheKey
+  private case class ProjectAuthCacheKey(userEmail: WorkbenchEmail, userToken: OAuth2BearerToken, googleProject: GoogleProject, executionContext: ExecutionContext) extends CacheKey
```

Comment:
Yeah good catch here -- breaking the `UserInfo` into a `(email, token)` for the cache key seems more correct. I can make a separate PR to do that for the SamAuthProvider.

Reply:
Opened issue: https://github.com/DataBiosphere/leonardo/issues/451